Implements Gather operator for arbitrary axis, sharing the code with BatchGather.

mike9ant opened the pull request:
Summary:
This implements general Gather operator for arbitrary axis, sharing the code with BatchGather.
 - CPU gather & batch gather logic is now shared through caffe2::gather_helper, for any axis.
 - Shared CUDA kernel moved to gather_op.cuh, for any axis.
 - Gradients of axis > 0 delegate to BatchGatherGradientOp which now has axis argument.
 - BatchGatherOp doc strings updated to have correct rank (q + (r -1)) and output.
 - Added tests for axis == 2.

TBD: Should we support negative index? Original GatherOp had it, but is not in docs and other places.

Differential Revision: D12983815

dzhulgakov commented on caffe2/operators/batch_gather_ops.cc:
```diff
@@ -12,24 +12,20 @@ OPERATOR_SCHEMA(BatchGather)
     .SetDoc(R"DOC(
 Batch gather operation, first dimension in DATA is the batch size.
 Given DATA tensor of rank r >= 2, and INDICES tensor of rank q >= 1, gather
 entries of the outer-most dimension of DATA indexed by INDICES, and concatenate
```
it's not outer-most dimension any more (this description was wrong before)

mike7ant replied:
will update!

dzhulgakov commented on caffe2/operators/batch_gather_ops.cc:
```diff
@@ -41,14 +37,14 @@ them in an output tensor of rank (q - 1) + (r - 1).
-      [1.0, 2.4],
-      [2.3, 3.6],
-      [4.5, 1.2],
+      [[1.0, 2.4],],
```
I think it would make more  sense to make INDICES 1-d above (both examples are fine, the 1d one is more natural)

dzhulgakov commented on caffe2/operators/batch_gather_ops.cu:
```diff
@@ -118,34 +72,50 @@ bool BatchGatherGradientOp<CUDAContext>::DoRunWithType2() {
-  CAFFE_ENFORCE_EQ(data.dim(0), grad.dim(0), "batch sizes should be the same");
+  // ONNX allows negative axis to index from the back, valid range: [-r, r].
+  int axis = axis_;
+  if (axis < 0) {
```
there was some `canonical_axis_index_` function somewhere

dzhulgakov commented on caffe2/operators/gather_op.cc:
```diff
@@ -75,43 +75,66 @@ print("OUTPUT:\n", workspace.FetchBlob("OUTPUT"))
+
+    // TBD: This is misleading to use dense_gradient by default for axis 0
+    // and not othewise....
+    if (argsHelper.HasArgument("dense_gradient")) {
```
I don't think we should enforce dense_gradient flag to be present if axis > 0. Effectively one can interpret `dense_gradient` as `force_dense_gradient` whereas def behavior is 'sparse only iff axis==0'

mike7ant replied:
This doesn't enforce dense_gradient, just checks its consistency for axis > 0 because there is no sparse impl for it  (unless you don't think that's good?). I'll rephrase the message.

dzhulgakov commented on caffe2/operators/gather_op.h:
```diff
@@ -6,11 +6,133 @@
+      for (auto i = 0; i < N; ++i) {
+        auto idx = idxs[i];
+        if (idx < 0) {
+          idx = idx + src_indexing_axis_dim;
```
wait a sec - where does this wrapping come from? we should totally throw on out-of-bounds access. That's what the previous code did.

Note, that wrapping around is for **axis index**, not the index being looked up.

mike7ant replied:
The gather_op.h implementation specifically had index wrapping, and not just axis; this is why I adopted it in other places.  However, perhaps that was a mistake? I can remove it everywhere, as i don't see how it makes much sense (but wouldn't want some unexpected failure from changing operator).

Speaking of range checking, should we add it on Cuda side? Because currently CPU version checks bounds with CAFFE_ENFORCE but CUDA kernels do nothing, which probably means out of bounds access.

dzhulgakov commented on caffe2/operators/batch_gather_ops.h:
```diff
@@ -4,6 +4,8 @@
+#include "caffe2/operators/gather_op.h"
 
 namespace caffe2 {
 
```
so the only difference between the functions right now is the default value for `axis`, you could template the entire class on `int default_axis` and register it twice

mike7ant replied:
Hmm, you mean get rid of batch_gather_op.h/cpp and reuse the class for both registrations? I've considered merging the sources... 

Actually, is there a way to dynamically check within the gather and gradient op what registration they come from, so as to reuse the code without a template?

dzhulgakov reviewed (commented):
Looking close, some comments

dzhulgakov reviewed (approved):
Looks good modulo "this" concern. I think it's good to go - but would be nice to actually do it without exposing Operator internals additionally (feel free to do so in a separate PR)

mike7ant commented on caffe2/operators/batch_gather_ops.cu:
```diff
@@ -37,58 +16,34 @@ bool BatchGatherOp<CUDAContext>::RunOnDevice() {
-  return true;
+  // BatchGather is a special-case of Gather with Axis = 1, wrap = false.
+  return gather_helper::gather_impl_cuda<TInd>(
+      this, DATA, INDICES, 0, 1, false);
```
I don't like it either but that's what I have to do if we want to share setup code and use Output(index, shape, datatype) instead of Resize. Else we'd have to copy/paste that setup code... so i think it's better to do it this was.

Commit:
Implements Gather operator for arbitrary axis, sharing the code with BatchGather. (#13756)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13756

This implements general Gather operator for arbitrary axis, sharing the code with BatchGather.
 - CPU gather & batch gather logic is now shared through caffe2::gather_helper, for any axis.
 - Shared CUDA kernel moved to gather_op.cuh, for any axis.
 - Gradients of axis > 0 delegate to BatchGatherGradientOp which now has axis argument.
 - BatchGatherOp doc strings updated to have correct rank (q + (r -1)) and output.
 - Added tests for axis == 2.

GatherOp supports index wrapping for axis == 0 by default, which was earlier for ONNX.
This diff also extends it to work in Cuda kernel. Added "wrap_indices" argument which specifies
wheather this wrapping should be done; set it to true if you'd like wrapping for any axis.

TBD: Update gradients to support negative indices (separate diff).
TBD: Once we have operator versioning, we'd like to update GatherOp to NOT support axis 0 wrapping
by default, but rather do it only if wrap_indices is set.

Differential Revision: D12983815

fbshipit-source-id: eba1b93293b8cf3569605cf915eef1a04858434f
